### ISSUE TITLE
Nested multiset record mapping fails at the 3rd level of nesting

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12155]
Given:

```java
static record NestA (Integer a, List<NestB> b) {}
static record NestB (Integer b, List<NestC> c) {}
static record NestC (Integer c, List<NestD> d) {}
static record NestD (Integer d, List<NestE> e) {}
static record NestE (Integer e) {}
```

Then this execution and mapping:

```java
var a =
ctx.select(
        val(1),
        multiset(select(
            val(2),
            multiset(select(
                val(3),
                multiset(select(
                    val(4),
                    multiset(select(val(5))).convertFrom(r -> r.map(mapping(NestE::new)))
                )).convertFrom(r -> r.map(mapping(NestD::new)))
            )).convertFrom(r -> r.map(mapping(NestC::new)))
        )).convertFrom(r -> r.map(mapping(NestB::new)))
    )
   .fetch(mapping(NestA::new));

System.out.println(a);
```

Produces an invalid result, both for the XML and JSON serialisation:

```
+----+---------------------------------------+
|   1|multiset                               |
+----+---------------------------------------+
|   1|[NestB[b=2, c=[NestC[c=null, d=null]]]]|
+----+---------------------------------------+
```

The following nesting without record mapping works correctly:

```java
println(ctx.select(
        val(1),
        multiset(select(
            val(2),
            multiset(select(
                val(3),
                multiset(select(
                    val(4),
                    multiset(select(val(5)))
                ))
            ))
        ))
    )
   .fetch()
   .formatJSON(new JSONFormat().format(true).header(false).recordFormat(RecordFormat.OBJECT)));
```

The output is:

```
[
  {
    "1": 1,
    "multiset": [
      {
        "2": 2,
        "multiset": [
          {
            "3": 3,
            "multiset": [
              {
                "4": 4,
                "multiset": [
                  { "5": 5 }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
]
```

And so does the executed query:

```sql
-- XML
select
  1,
  (
    select xmlelement(
      name result,
      xmlagg(xmlelement(
        name record,
        xmlelement(
          name "2",
          "v0"
        ),
        xmlelement(
          name "multiset",
          "v1"
        )
      ))
    )
    from (
      select
        2 as "v0",
        (
          select xmlelement(
            name result,
            xmlagg(xmlelement(
              name record,
              xmlelement(
                name "3",
                "v0"
              ),
              xmlelement(
                name "multiset",
                "v1"
              )
            ))
          )
          from (
            select
              3 as "v0",
              (
                select xmlelement(
                  name result,
                  xmlagg(xmlelement(
                    name record,
                    xmlelement(
                      name "4",
                      "v0"
                    ),
                    xmlelement(
                      name "multiset",
                      "v1"
                    )
                  ))
                )
                from (
                  select
                    4 as "v0",
                    (
                      select xmlelement(
                        name result,
                        xmlagg(xmlelement(
                          name record,
                          xmlelement(
                            name "5",
                            "v0"
                          )
                        ))
                      )
                      from (
                        select 5 as "v0"
                      ) as "t"
                    ) as "v1"
                ) as "t"
              ) as "v1"
          ) as "t"
        ) as "v1"
    ) as "t"
  )
-- JSON
select
  1,
  (
    select coalesce(
      jsonb_agg(jsonb_build_array("v0", "v1")),
      jsonb_build_array()
    )
    from (
      select
        2 as "v0",
        (
          select coalesce(
            jsonb_agg(jsonb_build_array("v0", "v1")),
            jsonb_build_array()
          )
          from (
            select
              3 as "v0",
              (
                select coalesce(
                  jsonb_agg(jsonb_build_array("v0", "v1")),
                  jsonb_build_array()
                )
                from (
                  select
                    4 as "v0",
                    (
                      select coalesce(
                        jsonb_agg(jsonb_build_array("v0")),
                        jsonb_build_array()
                      )
                      from (
                        select 5 as "v0"
                      ) as "t"
                    ) as "v1"
                ) as "t"
              ) as "v1"
          ) as "t"
        ) as "v1"
    ) as "t"
  )
```

Producing:

**XML**

```
|?column?|xmlelement|
|--------|----------|
|1       |[XML]     |
```

With XML being:

```xml
<result>
	<record>
		<_x0032_>2</_x0032_>
		<multiset>
			<result>
				<record>
					<_x0033_>3</_x0033_>
					<multiset>
						<result>
							<record>
								<_x0034_>4</_x0034_>
								<multiset>
									<result>
										<record>
											<_x0035_>5</_x0035_>
										</record>
									</result>
								</multiset>
							</record>
						</result>
					</multiset>
				</record>
			</result>
		</multiset>
	</record>
</result>
```

**JSON**

```
|?column?|coalesce                  |
|--------|--------------------------|
|1       |[[2, [[3, [[4, [[5]]]]]]]]|
```

**Comment by lukaseder** [comment]
The problem is in `ConvertedDataType`:

```java
    @Override
    public final U convert(Object object) {
        if (getConverter().toType().isInstance(object))
            return (U) object;

        // [#3200] Try to convert arbitrary objects to T
        else
            return ((Converter<T, U>) getConverter()).from(delegate.convert(object));
    }
```

We're converting back and forth between `Result<Record[N]>` and `List<UserType>`, only we don't actually know how to convert a `List<UserType>` back into a `Result<Record[N]>` anymore for a re-conversion.

As it stands now, this would fix the conversion issue for the XML serialisation:

```java
    @Override
    public final U convert(Object object) {
        if (getConverter().toType().isInstance(object))
            return (U) object;

        else if (delegate.isMultiset() && !(object instanceof Result))
            return (U) object;

        // [#3200] Try to convert arbitrary objects to T
        else
            return ((Converter<T, U>) getConverter()).from(delegate.convert(object));
    }
```

But not for the JSON one. To be further investigated

**Comment by lukaseder** [comment]
The real underlying problem here is that we never remember whether we've already converted between `T` and `U`, or if we still have to do this. This produces some performance issues in the past for slow conversions (and this one can be considered slow as well!)

I'll investigate this in a separate, future issue, to see whether we can optimise conversions in a way to remember whether it is still required.

**Comment by lukaseder** [comment]
The problem with the `JSONReader` is that it doesn't recurse like the `XMLHandler`, which is a `SAXHandler`, which recurses by design. It should recurse into its nested `List` and `Map` objects to convert the inner-most data structures first, bubbling back up.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.1 (https://github.com/jOOQ/jOOQ/issues/12156)
